File: src/main/rust/periodic_table.rs
use crate::errors::{ChemikazeError, ErrorKind};
use crate::util::bytes_to_string;

/// They are roughly sorted by popularity in organic chemistry. Well, at least the first elements
/// are. Doesn't contain elements that would never be used in organic chemistry.
pub const EARTH_SYMBOLS: [&str; 85] = [
    "H", "C", "O", "N", "P", "F", "S", "Br", "Cl", "Na", "Li", "Fe", "K", "Ca", "Mg", "Ni", "Al",
    "Pd", "Sc", "V", "Cu", "Cr", "Mn", "Co", "Zn", "Ga", "Ge", "As", "Se", "Ti", "Si", "Be", "B",
    "Kr", "Rb", "Sr", "Y", "Zr", "Nb", "Mo", "Ru", "Rh", "Ag", "Cd", "In", "Sn", "Sb", "Te", "I",
    "Xe", "Cs", "Ba", "La", "Ce", "Pr", "Nd", "Sm", "Eu", "Gd", "Tb", "Dy", "Ho", "Er", "Tm", "Yb",
    "Lu", "Hf", "Ta", "Tc", "W", "Re", "Os", "Ir", "Pt", "Au", "Hg", "Tl", "Pb", "Bi", "Th", "Pa",
    "U", "He", "Ne", "Ar",
];
pub const EARTH_ELEMENT_CNT: usize = EARTH_SYMBOLS.len();
/// Same as EARTH_SYMBOLS, but keep symbols as bytes so that we don't have to turn them into
/// String on each lookup. Each symbol is 2 bytes: for 1-symbol elements like H, the 2nd byte is 0.
const EARTH_SYMBOLS_AS_BYTES: [u8; EARTH_ELEMENT_CNT*2] = build_symbols_as_bytes();
/// The size of {@link #ELEMENTHASH_TO_ELEMENT} hash table
const INDEX_BUCKET_CNT: usize = 512;
const INDEX_HASH_MASK: usize = INDEX_BUCKET_CNT - 1;
const ELEMENTHASH_TO_ELEMENT: [u8; INDEX_BUCKET_CNT] = build_index();

/// `symbol` - "H", "Na", etc. Only the elements that actually exist on the Earth are used,
///           see `EARTH_SYMBOLS` array.
pub fn get_element_by_symbol_str(symbol: &str) -> Result<u8, ChemikazeError> {
    let ascii = symbol.as_bytes();
    let mut bytes: [u8; 2] = [0; 2];
    bytes[0] = ascii[0];
    if ascii.len() > 1 {
        bytes[1] = ascii[1];
    }
    get_element_by_symbol_bytes(bytes)
}
/// `bytes` represent symbol ASCII (like ['H', 'e']). For 1-byte symbols: ['H', 0].
pub fn get_element_by_symbol_bytes(bytes: [u8; 2]) -> Result<u8, ChemikazeError> {
    let element = ELEMENTHASH_TO_ELEMENT[hash(bytes)];
    let i = (element * 2) as usize; // Java impl doesn't need to multiply here
    if EARTH_SYMBOLS_AS_BYTES[i] != bytes[0] || EARTH_SYMBOLS_AS_BYTES[i+1] != bytes[1] {
        let mut element_str = bytes_to_string(&bytes);
        if bytes[1] == 0 {
            element_str = bytes_to_string(&bytes[0..1]);
        }
        return Err(ChemikazeError {
            msg: String::from(format!("Unknown chemical symbol: {element_str}")),
            kind: ErrorKind::UnknownElement
        });
    }
    Ok(element)
}
pub const fn hash(symbol: [u8; 2]) -> usize {
    // Ran an experiment, and 277 is one of few multipliers that gave no collisions in 512-sized
    // hash table. Couldn't achieve the same with subtractions or shifts, no matter the order of b0
    // and b1.
    ((symbol[0] as usize * 277) ^ symbol[1] as usize) & INDEX_HASH_MASK
}

const fn build_index() ->  [u8; INDEX_BUCKET_CNT] {
    // a precaution to see if we made an error in the hash()
    let mut taken_buckets = [false; INDEX_BUCKET_CNT];
    let mut result = [0; INDEX_BUCKET_CNT];
    let mut i = 0;
    while i < EARTH_SYMBOLS.len() {
        let bucket = hash([EARTH_SYMBOLS_AS_BYTES[i*2], EARTH_SYMBOLS_AS_BYTES[i*2+1]]);
        if taken_buckets[i] {
            panic!("Wrong hash function for the Symbol HashTable: 2 symbols were mapped to the same bucket")
        }
        taken_buckets[i] = true;
        result[bucket] = i as u8;
        i += 1;
    }
    result
}
const fn build_symbols_as_bytes() -> [u8; EARTH_ELEMENT_CNT*2] {
    let mut symbol_bytes: [u8; EARTH_ELEMENT_CNT*2] = [0u8; EARTH_ELEMENT_CNT*2];
    let mut i = 0;
    while i < EARTH_ELEMENT_CNT {
        let b = EARTH_SYMBOLS[i].as_bytes();
        let j = i * 2;
        symbol_bytes[j] = b[0];
        if b.len() > 1 {
            symbol_bytes[j + 1] = b[1];
        }
        i += 1;
    }
    symbol_bytes
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn returns_element_by_its_symbol_str() {
        assert_eq!(0u8, get_element_by_symbol_str("H").unwrap());
        assert_eq!(1u8, get_element_by_symbol_str("C").unwrap());
        assert_eq!(9u8, get_element_by_symbol_str("Na").unwrap());
    }
    #[test]
    fn returns_element_by_its_symbol_bytes() {
        assert_eq!(0u8, get_element_by_symbol_bytes(['H' as u8, 0]).unwrap());
        assert_eq!(82u8, get_element_by_symbol_bytes(['H' as u8, 'e' as u8]).unwrap());
    }
}

File: src/main/c/periodic_table.c
#include "periodic_table.h"

#include <stdio.h>
#include <stdlib.h>

#include "error.h"

#define INDEX_BUCKET_CNT 512
#define INDEX_HASH_MASK (INDEX_BUCKET_CNT-1)

unsigned hash(const char symbol[static 2]) {
	// Ran an experiment, and 277 is one of few multipliers that gave no collisions in 512-sized hash table.
	// Couldn't achieve the same with subtractions or shifts, no matter the order of b0 and b1.
	//
	// Need to try gperf in the future: https://www.gnu.org/software/gperf/manual/gperf.html#Description
	// ReSharper disable once CppRedundantParentheses
	return ((symbol[0] * 277) ^ symbol[1]) & INDEX_HASH_MASK;
}

// precomputed hash table of `hash(EARTH_SYMBOL_PADDED) -> ChemElement`. See main() for the precomputing.
constexpr ChemElement ELEMENTHASH_TO_ELEMENT[INDEX_BUCKET_CNT] = {
	[488]=0,[127]=1,[379]=2,[102]=3,[144]=4,[446]=5,[463]=6,[280]=7,[19]=8,[7]=9,[85]=10,[475]=11,[295]=12,[30]=13,
	[310]=14,[15]=15,[57]=16,[244]=17,[428]=18,[270]=19,[10]=20,[13]=21,[319]=22,[16]=23,[268]=24,[178]=25,[182]=26,
	[38]=27,[426]=28,[141]=29,[422]=30,[271]=31,[362]=32,[341]=33,[216]=34,[445]=35,[77]=36,[272]=37,[4]=38,[318]=39,
	[207]=40,[210]=41,[50]=42,[27]=43,[147]=44,[417]=45,[429]=46,[129]=47,[253]=48,[349]=49,[12]=50,[267]=51,[93]=52,
	[26]=53,[226]=54,[2]=55,[418]=56,[220]=57,[183]=58,[134]=59,[493]=60,[391]=61,[219]=62,[137]=63,[47]=64,[73]=65,
	[398]=66,[133]=67,[135]=68,[35]=69,[223]=70,[264]=71,[143]=72,[228]=73,[32]=74,[399]=75,[136]=76,[242]=77,[259]=78,
	[140]=79,[241]=80,[505]=81,[397]=82,[3]=83,[39]=84,
};

ChemElement ptable_getElementBySymbol(char symbol[static 2]) {
	ChemElement e = ELEMENTHASH_TO_ELEMENT[hash(symbol)];
	if (EARTH_SYMBOLS[e][0] != symbol[0] || EARTH_SYMBOLS[e][1] != symbol[1])
		return INVALID_CHEM_ELEMENT;
	return e;
}

[[maybe_unused]]
static void precomputeElementHashTable() {
	// Precompute values for ELEMENTHASH_TO_ELEMENT
	for (ChemElement i = 0; i < EARTH_ELEMENT_CNT; i++)
		printf("[%d]=%d,", hash((char[]) {EARTH_SYMBOLS[i][0], EARTH_SYMBOLS[i][1]}), i);
}